File: chap05/cat.go
package main

import (
	"fmt"
	"io"
	"log"
	"os"
)

const BufferSize = 2048

func main() {
	argc := len(os.Args)

	if argc < 2 {
		_, _ = fmt.Fprintf(os.Stderr, "%s: file name not given\n", os.Args[0])
		os.Exit(1)
	}

	for _, path := range os.Args[1:] {
		doCat(path)
	}

}


func doCat(path string) {
	buf := make([]byte, BufferSize)

	file, err := os.Open(path)
	defer file.Close()

	if err != nil {
		log.Fatal(err)
	}

	for {
		// Goなら読み込んだバイト数じゃなくてerrで判定するよな
		_, err := file.Read(buf)

		if err == io.EOF {
			break
		}

		if _, err := io.WriteString(os.Stdout, string(buf)); err != nil {
			log.Fatal(err)
		}

		if err != nil {
			log.Fatal(err)
		}

	}
}

File: chap05/cat.py
import os
import sys


def main():
    argv = sys.argv
    argc = len(argv)

    if argc < 2:
        print(f'{argv[0]}: file name not given', file=sys.stderr)
        exit(1)

    for file_path in argv[1:]:
        do_cat(file_path)


def do_cat(file_path: str) -> None:
    buffer_size = 6

    fd = os.open(file_path, os.O_RDONLY)

    while True:
        byte = os.read(fd, buffer_size)

        if len(byte) == 0:
            break

        os.write(sys.stdout.fileno(), byte)


if __name__ == '__main__':
    main()
